### PR TITLE
support row expression.

### PR DESCRIPTION
**Pull request by hanfei19910905**
e.g. select * from src where (a, b) = (1, 1)

### CODE REVIEW

File: plan/expression_rewriter.go [hunk 1]
```diff
@@ -38,11 +47,59 @@ type expressionRewriter struct {
...
-		return nil, nil
+func constructBinaryOpFunction(l *rowExpr, r *rowExpr, op opcode.Op) (*expression.ScalarFunction, error) {
+	if l.expr != nil && r.expr != nil {
+		return expression.NewFunction(opcode.Ops[op], []expression.Expression{l.expr, r.expr}, types.NewFieldType(mysql.TypeTiny))
+	} else if l.expr != nil {
+		return nil, errors.New("Operand should contain 1 column(s)")
+	} else if r.expr != nil || len(l.rows) != len(r.rows) {
+		return nil, errors.Errorf("Operand should contain %d column(s)", len(l.rows))
+	}
+	result, err := constructBinaryOpFunction(l.rows[0], r.rows[0], op)
```
Comment (coocood) [inline plan/expression_rewriter.go]: We can append result to a slice, and use `composeConditions` function to compose them.

File: plan/expression_rewriter.go [hunk 2]
```diff
@@ -108,14 +177,37 @@ func (er *expressionRewriter) Enter(inNode ast.Node) (retNode ast.Node, skipChil
 		np = er.b.buildMaxOneRow(np)
 		if np.IsCorrelated() {
 			er.p = er.b.buildApply(er.p, np, outerSchema)
-			er.ctxStack = append(er.ctxStack, er.p.GetSchema()[len(er.p.GetSchema())-1])
+			newExpr := make([]*rowExpr, 0, len(np.GetSchema()))
```
Comment (coocood) [inline plan/expression_rewriter.go]: 'newExpr' doesn't like a slice name.

File: plan/expression_rewriter.go [hunk 3]
```diff
@@ -108,14 +177,37 @@ func (er *expressionRewriter) Enter(inNode ast.Node) (retNode ast.Node, skipChil
...
+				er.err = errors.Trace(err)
+				return retNode, true
+			}
 			d, err := EvalSubquery(np, er.b.is, er.b.ctx)
 			if err != nil {
 				er.err = errors.Trace(err)
 				return retNode, true
 			}
-			er.ctxStack = append(er.ctxStack, &expression.Constant{Value: d[0], RetType: np.GetSchema()[0].GetType()})
+			newExpr := make([]*rowExpr, 0, len(np.GetSchema()))
```
Comment (coocood) [inline plan/expression_rewriter.go]: same as above.

File: plan/expression_rewriter.go [hunk 4]
```diff
@@ -124,16 +216,30 @@ func (er *expressionRewriter) Enter(inNode ast.Node) (retNode ast.Node, skipChil
...
 	}
 	switch v := inNode.(type) {
 	case *ast.AggregateFuncExpr:
+	case *ast.RowExpr:
+		length := len(v.Values)
+		rows := make([]*rowExpr, 0, length)
+		for i := stkLen - length; i < stkLen; i++ {
+			rows = append(rows, er.ctxStack[i])
+		}
+		er.ctxStack = er.ctxStack[:len(er.ctxStack)-length]
```
Comment (coocood) [inline plan/expression_rewriter.go]: len(er.ctxStack) -> stkLen

File: plan/expression_rewriter.go [hunk 5]
```diff
@@ -124,16 +216,30 @@ func (er *expressionRewriter) Enter(inNode ast.Node) (retNode ast.Node, skipChil
...
+
 	case *ast.FuncCallExpr:
 		function := &expression.ScalarFunction{FuncName: v.FnName}
-		for i := length - len(v.Args); i < length; i++ {
-			function.Args = append(function.Args, er.ctxStack[i])
+		for i := stkLen - len(v.Args); i < stkLen; i++ {
+			if er.ctxStack[i].expr != nil {
+				function.Args = append(function.Args, er.ctxStack[i].expr)
+			} else {
+				er.err = errors.New("Operand should contain 1 column(s)")
```
Comment (coocood) [inline plan/expression_rewriter.go]: return error first, then we don't need 'else'.

File: plan/newplanbuilder.go [hunk 6]
```diff
@@ -459,6 +459,13 @@ type havingAndOrderbyResolver struct {
 }
 
 func (e *havingAndOrderbyResolver) addProjectionExpr(v *ast.ColumnNameExpr, projCol *expression.Column) {
+	// avoid to append same column repeatly.
```
Comment (zimulala) [inline plan/newplanbuilder.go]: The first letter of the sentence is capitalized.

File: plan/expression_rewriter.go [hunk 7]
```diff
@@ -108,14 +167,37 @@ func (er *expressionRewriter) Enter(inNode ast.Node) (retNode ast.Node, skipChil
 		np = er.b.buildMaxOneRow(np)
 		if np.IsCorrelated() {
 			er.p = er.b.buildApply(er.p, np, outerSchema)
-			er.ctxStack = append(er.ctxStack, er.p.GetSchema()[len(er.p.GetSchema())-1])
+			newCols := make([]*rowExpr, 0, len(np.GetSchema()))
```
Comment (coocood) [inline plan/expression_rewriter.go]:
```
if len(np.GetSchema()) == 1 {
    er.ctxStack = append(er.ctxStack, np.GetSchema()[1])
}
```

File: plan/expression_rewriter.go [hunk 8]
```diff
@@ -47,46 +42,41 @@ type expressionRewriter struct {
...
 }
 
-func constructBinaryOpFunction(l *rowExpr, r *rowExpr, op string) (ret *expression.ScalarFunction, err error) {
-	if l.expr != nil && r.expr != nil {
-		return expression.NewFunction(op, types.NewFieldType(mysql.TypeTiny), l.expr, r.expr)
-	} else if l.expr != nil {
-		return nil, errors.New("Operand should contain 1 column(s)")
-	} else if r.expr != nil || len(l.rows) != len(r.rows) {
-		return nil, errors.Errorf("Operand should contain %d column(s)", len(l.rows))
+func checkRowLen(e expression.Expression) int {
```
Comment (coocood) [inline plan/expression_rewriter.go]: `checkXXX` should return boolean value,
rename to `getRowLen` is better.

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**